### PR TITLE
Fixed Bugs in Utility Function of RUBI

### PR DESCRIPTION
**Pull request by ashishkg0022**
This pr contains the following change:
1. `Dist` in utility function was not defined properly
2. Added `IntegralFreeQ` 
3. a minor bug in `SimpHelp` where it should have been `w.args`
4. `NumberQ`

But the test is failing because, there is also an error while calling `Condition` in `_FixSimplfy()`
`Condition(FixSimplify(Mul(w, Pow(S('q'), m), Pow(Add(c, d), Add(m, n)))))`
Only single argument is passed , but two are required. Can I get some help to resolve this ?

### CODE REVIEW

**Comment by ashishkg0022** [comment]
ping @Upabjojr

**Comment by Upabjojr** [comment]
Hi, are you importing the implementation from the RUBI code?

**Comment by ashishkg0022** [comment]
Yes, I tried to do similar to implementaion in the RUBI code. I will also write few test case for `Dist` and `IntegralFreeQ`, once I resolve error with  `Condition`

**Comment by Upabjojr** [comment]
OK. Do you have a Mathematica license to compare the results?

**Comment by ashishkg0022** [comment]
No , they give 15 days free trial, I am using that.

**Comment by Upabjojr** [comment]
I see. There are probably lots of bugs in the utility functions. RUBI also gives wrong results in most cases.

**Comment by ashishkg0022** [comment]
I have found few more bugs, I will try to resolve them. Can you help me in this
```
 File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.1.2.dev0-py3.6.egg/sympy/integrals/rubi/utility_function.py", line 6274, in <lambda>
    rule8 = ReplacementRule(pattern8, lambda a, n, c, b, d, m, w : With(List(Set(S('q'), Simplify(Mul(b, Pow(d, S(-1)))))), Condition(FixSimplify(Mul(w, Pow(S('q'), m), Pow(Add(c, d), Add(m, n)))))))
TypeError: Condition() missing 1 required positional argument: 'c'
```
Are these code written using parse_tools ?

**Comment by Upabjojr** [comment]
> Are these code written using parse_tools ?

Yes, that output looks like the product of automatic tools we used in GSoC 2017.

**Comment by ashishkg0022** [comment]
How can I resolve it ?

**Comment by Upabjojr** [comment]
> How can I resolve it ?

Did you figure out how to use the tools?

**Comment by ashishkg0022** [comment]
No

**Comment by Upabjojr** [comment]
The parsed Mathematica expressions are contained here:

https://github.com/Upabjojr/RUBI_integration_rules

The parse tools should operate on one of those files.

**Comment by ashishkg0022** [comment]
Ok, I will try to resolve it . thanks !

**Comment by ashishkg0022** [comment]
@Upabjojr  , I think in this pr also `Dist`  is not fully implemented. for full `Dist` we have to implement an object that behaves like that under addition or subtraction. So I just implemented the normal part.
```
Dist[u_,v_,x_] := 
  If[u===1,
    v,
  If[u===0,
    Print["*** Warning ***:  Dist[0,",v,",",x,"]"];
    0,
  If[NumericFactor[u]<0 && NumericFactor[-u]>0,
    -Dist[-u,v,x],
  If[SumQ[v],
    Map[Function[Dist[u,#,x]],v],
  If[IntegralFreeQ[v],
(*  Simp[Simp[u,x]*v,x], *)
    Simp[u*v,x],
(*If[ShowSteps=!=True,
    Simp[u*v,x],
  Module[{w=Simp[u,x]},
  If[w=!=u,
    Dist[w,v,x], *)
  With[{w=Simp[u*x^2,x]/x^2},
  If[w=!=u && FreeQ[w,x] && w===Simp[w,x] && w===Simp[w*x^2,x]/x^2,
    Dist[w,v,x],
  If[ShowSteps=!=True,
    Simp[u*v,x],
  Defer[Dist][u,v,x]]]]]]]]]
```
this is the simple part from RUBI.
Can you suggest  any way for proper implementation of `Dist` ?

**Comment by Upabjojr** [comment]
Currently in SymPy there is an _op_priority mechanism to determine which `__add__` or `__radd__` or other operators to use.

**Comment by Upabjojr** [comment]
The default op priority is 10, if it's set higher the operator methods will prevail over Add and Mul.

**Comment by ashishkg0022** [comment]
Is the simple implementation of `Dist` as in this pr not sufficient ? I think , it will handle the cases like 
```
Dist /: Dist[u_,v_,x_]+Dist[w_,v_,x_] := 
  If[EqQ[u+w,0],
    0,
  Dist[u+w,v,x]]

Dist /: Dist[u_,v_,x_]-Dist[w_,v_,x_] := 
  If[EqQ[u-w,0],
    0,
  Dist[u-w,v,x]]
```

**Comment by Upabjojr** [comment]
It could. Did you compare Mathematica's output? We need some tests for every new feature added.

File: sympy/integrals/rubi/utility_function.py [hunk 1]
```diff
@@ -6062,7 +6062,7 @@ def SimpHelp(u, x):
                 w = i + w
         v = SmartSimplify(v)
         if SumQ(w):
-            w = Add(*[SimpHelp(i, x) for i in w])
+            w = Add(*[SimpHelp(i, x) for i in w.args])
```
Comment (Upabjojr) [inline sympy/integrals/rubi/utility_function.py]: Why this change?

**Comment by ashishkg0022** [comment]
Yes, I compared few outputs , they were coming right. I will add test for the changes.

File: sympy/integrals/rubi/utility_function.py (hunk 1 above)
Comment (ashishkg0022) [reply sympy/integrals/rubi/utility_function.py]: I think that is an error, because in the above part of code , we are doing `w = i + w`. So 
`for i in w :` will give error as 
```
w = Add(*[SimpHelp(i, x) for i in w])
TypeError: 'Add' object is not iterable
```

**Comment by ashishkg0022** [comment]
@Upabjojr , I have added tests. Is it fine ?

File: sympy/integrals/rubi/utility_function.py [hunk 2]
```diff
@@ -6678,8 +6680,27 @@ def PolyLog(n, p, z=None):
 def D(f, x):
     return f.diff(x)
 
+def IntegralFreeQ(u):
```
Comment (Upabjojr) [inline sympy/integrals/rubi/utility_function.py]: I notice that this utility function does not exist in the RUBI files. I would not recommend to introduce new utility functions. SymPy already has its own. Besides, it appears that this utility function is only needed in one place.

Comment (Upabjojr) [reply sympy/integrals/rubi/utility_function.py]: Sorry, was using an old version of RUBI.

**Comment by ashishkg0022** [comment]
@Upabjojr 
> was using an old version of RUBI.
Do we need to update the functions with newer versions ? Or should we fix to a version ?

**Comment by Upabjojr** [comment]
> Do we need to update the functions with newer versions ? Or should we fix to a version ?

Updating makes everything more complicated. Eventually we will need to update to the newer version.

I would prefer to focus the attention on getting the integrator to work. Many integrals are still wrong.

**Comment by ashishkg0022** [comment]
ok, I will work on it. Thanks! 
Can we discuss the works that must be done to get this integrator working?

**Comment by Upabjojr** [comment]
The structure of RUBI has been ported. The problem is that there are many bugs.

**Comment by ashishkg0022** [comment]
ok, I will try to find and fix.

**Comment by Upabjojr** [comment]
There are test cases on the RUBI website. They should be ported and tried. Some will give the wrong result, some others will raise exceptions. Lots of things to fix.

**Comment by parsoyaarihant** [comment]
We should also try to remove `exp()` and just use `pow()` : https://github.com/sympy/sympy/pull/13211

This is needed for matching the SymPy expression directly with the matcher. Else, we need to preprocess the expression to replace `exp()` with `pow()`.

**Comment by ashishkg0022** [comment]
@Upabjojr I was testing the results , but trial version pf mathematica has ended. Is there any way I could get it without paying ? Or do I have to buy ?

**Comment by Upabjojr** [comment]
There is Mathics, but it's quite incomplete. Did you check whether your institution has software support?

**Comment by ashishkg0022** [comment]
I checked. My institution has no license for Mathematica.
